### ISSUE TITLE
Websocket error in v2.0

### ISSUE THREAD

**Issue by ronag** [deepstreamIO/deepstream.io-client-js#266]
```
07:15:34.346Z ERROR asset: Deepstream Error. (error="Two connections heartbeats missed successively", event=connectionError, topic=C)
node_modules/ws/lib/WebSocket.js:219
    else throw new Error('not opened');
         ^

Error: not opened
    at WebSocket.send (node_modules/ws/lib/WebSocket.js:219:16)
    at Connection._handleConnectionResponse (node_modules/deepstream.io-client-js/src/message/connection.js:386:18)
    at Connection._onMessage (node_modules/deepstream.io-client-js/src/message/connection.js:349:9)
    at WebSocket.onMessage (node_modules/ws/lib/WebSocket.js:442:14)
    at emitTwo (events.js:106:13)
    at WebSocket.emit (events.js:191:7)
    at Receiver.ontext (node_modules/ws/lib/WebSocket.js:841:10)
    at node_modules/ws/lib/Receiver.js:536:18
    at Receiver.applyExtensions (node_modules/ws/lib/Receiver.js:371:5)
    at node_modules/ws/lib/Receiver.js:508:14
[nodemon] app crashed - waiting for file changes before starting...
```

**Comment by yasserf** [comment]
I saw this issue in ws in a few of our test frameworks, I tracked it down to an odd issue with ws not propagating the close event before changing its state to closed.

When you raised issues against master/unreleased versions can you please mention the version? Having an issue just titled  `crashes` sounds a bit intense.

**Comment by ronag** [comment]
@yasserf: Yes sorry about that. Was in a hurry.

**Comment by ronag** [comment]
@yasserf: Is this something that needs to be fixed in WS or DS?

**Comment by yasserf** [comment]
Not entirely sure yet tbh, I have a fix in DS that checks for the connection state on the websocket but is a workaround.

**Comment by ronag** [comment]
@yasserf: send can throw. https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#Exceptions_thrown_2

**Comment by yasserf** [comment]
send can throw when the connection state is not open, which is something we should get notified about via ws.

A better solution would be to check if the connection ready state is open rather than via try catch

```
if (this._endpoint.readyState !== WebSocket.OPEN) {
 // do something like log an error or queue message
}
```

**Comment by ronag** [comment]
> which is something we should get notified about via ws.

Are you sure it has to notify synchronously before you are able to call send?

**Comment by yasserf** [comment]
Nope, which is why we are going to add the above guard. I just want to avoid having a try catch since they tend to perform quite poorly.

> which is something we should get notified about via ws.

true, engine.io mentions this [here](https://github.com/socketio/engine.io-client/blob/9cfc626bf7c5ccf9531549b36515551a96352fa8/lib/transports/websocket.js):

```
        // Sometimes the websocket has already been closed but the browser didn't
        // have a chance of informing us about it yet, in that case send will
        // throw an error
        try {
          if (self.usingBrowserWebSocket) {
            // TypeError is thrown when passing the second argument on Safari
            self.ws.send(data);
          } else {
            self.ws.send(data, opts);
          }
        } catch (e) {
          debug('websocket closed before onclose event');
        }
```

**Comment by ronag** [comment]
Which brings me to the next question. Is `readyState` guaranteed to be synchronously correct?

**Comment by yasserf** [comment]
According to the `ws` library, yes:

https://github.com/websockets/ws/blob/5a025dcb730427be1d696718003784b33babd696/lib/WebSocket.js

I have implemented it in some of our test clients and they stopped failing. I would rather put that in, and resort to a try catch if it doesn't work.

**Comment by ronag** [comment]
@yasserf @jdmnd Your fix does not fix the issue. Same error at https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/message/connection.js#L373

I would recommend having a `_send` method that performs the check.

**Comment by ronag** [comment]
https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/message/connection.js#L384
https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/message/connection.js#L230

**Comment by jdmnd** [comment]
Good catch, I'll take a look.

**Comment by ronag** [comment]
@jdmnd: What is the status on this?

**Comment by jdmnd** [comment]
@ronag I opened a new pull request with the changes you suggested. 
https://github.com/deepstreamIO/deepstream.io-client-js/pull/274
The check on this line remains so that messages aren't removed from the queue if the socket is closed:
https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/message/connection.js#L374
